HerPublicWebsite.UnitTests/BusinessLogic/QuestionnaireUpdaterTests.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System.Threading.Tasks;
+using FluentAssertions;
+using HerPublicWebsite.BusinessLogic;
+using HerPublicWebsite.BusinessLogic.ExternalServices.EpbEpc;
+using HerPublicWebsite.BusinessLogic.Models;
+using HerPublicWebsite.BusinessLogic.Services.EligiblePostcode;
+using Moq;
+using NUnit.Framework;
+
+namespace Tests.BusinessLogic;
+
+[TestFixture]
+public class QuestionnaireUpdaterTests
+{
+    private QuestionnaireUpdater underTest;
+    private Mock<IEpcApi> mockEpcApi;
+    private Mock<IEligiblePostcodeService> mockPostCodeService;
+    
+    
+    [SetUp]
+    public void Setup()
+    {
+        mockEpcApi = new Mock<IEpcApi>();
+        mockPostCodeService = new Mock<IEligiblePostcodeService>();
+        underTest = new QuestionnaireUpdater(mockEpcApi.Object, mockPostCodeService.Object);
+    }
+    
+    [Test]
+    public async Task UpdateAddressAsync_CalledWithUprn_GetsEpcDetails()
+    {
+        // Arrange
+        var questionnaire = new Questionnaire();
+        var address = new Address()
+        {
+            AddressLine1 = "line1",
+            County = "county",
+            Postcode = "ab1 2cd",
+            Uprn = "123456789012"
+        };
+        var epcDetails = new EpcDetails()
+        {
+            AddressLine1 = "epc line 1",
+        };
+        mockEpcApi.Setup(e => e.EpcFromUprnAsync("123456789012")).ReturnsAsync(epcDetails);
+        
+        // Act
+        var result = await underTest.UpdateAddressAsync(questionnaire, address);
+        
+        // Assert
+        mockEpcApi.Verify(e => e.EpcFromUprnAsync("123456789012"));
+        result.EpcDetails.Should().Be(epcDetails);
+        result.EpcDetailsAreCorrect.Should().BeNull();
+    }
+    
+    [Test]
+    public async Task UpdateAddressAsync_CalledWithoutUprn_ResetsEpcDetails()
+    {
+        // Arrange
+        var questionnaire = new Questionnaire
+        {
+            EpcDetails = new EpcDetails()
+        };
+        var address = new Address()
+        {
+            AddressLine1 = "line1",
+            County = "county",
+            Postcode = "ab1 2cd"
+        };
+
+        // Act
+        var result = await underTest.UpdateAddressAsync(questionnaire, address);
+        
+        // Assert
+        mockEpcApi.VerifyNoOtherCalls();
+        result.EpcDetails.Should().BeNull();
+        result.EpcDetailsAreCorrect.Should().BeNull();
+    }
+    
+    [TestCase(true)]
+    [TestCase(false)]
+    public async Task UpdateAddressAsync_WhenCalled_SetsLsoaStatusCorrectly(bool isEligible)
```

Comment:
Very nitpicky, but I was told once that you should avoid using words like "correctly" in test names, since it's a tautology! I think this could be something like `...SetsLsoaStatusToPostcodeEligibility` or something

Reply:
Fair, I find it more difficult to name tests that way when they have multiple test cases